### PR TITLE
add ability to use http proxy when making web requests

### LINKED ISSUE #241: [feature request] Add ability for HTTP Config Manager to take proxy options
I propose adding the following options to HTTP Config manager:

| name | default | description |
| --- | --- | --- |
| proxy_host | nil | To perform a web request behind a proxy, this is required. This can be a domain or IP address of the proxy server.
| proxy_port | nil | port number that the proxy server uses for client connections. |
| proxy_username | nil | username for proxy auth. |
| proxy_password | nil | password for proxy auth. |

Happy to submit a PR to add this functionality.

### PR DESCRIPTION
**Pull request by rloomba**
## Summary
- add ability to use a proxy server when making http requests
- modify`HTTPProjectConfigManager` and `EventDispatcher` to take in proxy config options

I am unable to use this library as all my web requests must be routed through a web proxy.

[HTTParty proxy config documentation](https://github.com/jnunemaker/httparty/blob/master/lib/httparty.rb#L37-L40)

## Issues
- Fixes #241

### CODE REVIEW

**Comment by rloomba** [comment]
@thomaszurkan-optimizely sounds good! Happy to do the refactor.

**Comment by jasonkarns** [comment]
Does it make sense to further entrench the httparty gem via this configuration? As I opened in #211 , I think it makes sense to eliminate the gem dependency entirely and just depend on Ruby Core/StdLib alone. (ie, net/http or open-uri) Both of the stdlib tools respect HTTP Proxy env vars out of the box and would require no configuration at all.

**Comment by rloomba** [comment]
I'm 100% onboard with #211.

I don't think we'll entrench this gem that much further If we extract the proxy config settings to a separate config object that @thomaszurkan-optimizely mentioned [above](https://github.com/optimizely/ruby-sdk/pull/242#pullrequestreview-394003898), couldn't that be used in whatever HTTP library we end up moving to?

I think respecting the HTTP Proxy env var is a good approach, but I also like having the option to explicitly config the proxy in application level code. What if we have multiple proxies that we'd like switch between?

@jasonkarns implementing #211 seems like a bigger chunk of work that I'm not sure I can lead, but  I'm happy to pitch in if there are some smaller pieces to be worked on or at least offer some code review.

**Comment by jasonkarns** [comment]
> I think respecting the HTTP Proxy env var is a good approach, but I also like having the option to explicitly config the proxy in application level code. What if we have multiple proxies that we'd like switch between?

Good call. The trick will be ensuring that we distinguish the options being unset (to allow env vars to affect), from them being intentionally set to nil (to ignore env vars and disable the proxy).

**Comment by thomaszurkan-optimizely** [comment]
Hi @rloomba,
We removed httparty and created a HttpUtils class to wrap our http calls.  Can you refactor your proxy settings here and we can merge it?  Otherwise, we will do it sometime in the next few sprints.  Thanks!

**Comment by rloomba** [comment]
@thomasgodwell-optimizely ah sorry for letting this drag on. Yes, happy to rebase master and make those changes sometime this week.

**Comment by rloomba** [comment]
@thomasgodwell-optimizely tooka pass at this refactor, let me know if you have any feedback.

**Comment by oakbani** [comment]
@msohailhussain  and @thomaszurkan-optimizely  Tested FSC Suite here https://github.com/optimizely/ruby-sdk/pull/261

File: .travis.yml [hunk 1]
```diff
@@ -43,7 +43,8 @@ jobs:
       install:
         - npm i -g markdown-spellcheck
       before_script:
-        - wget --quiet https://raw.githubusercontent.com/optimizely/mdspell-config/master/.spelling
+        # todo: change branch to master once merged.
+        - wget --quiet https://raw.githubusercontent.com/optimizely/mdspell-config/oakbani/ignore-for-ruby/.spelling
```
Comment (thomaszurkan-optimizely) [inline .travis.yml:47]: can we merge and fix this?

**Comment by thomaszurkan-optimizely** [comment]
This was move to PR #263 so that we could run FSC tests.  @rloomba thank you!  I made sure to include your credit in the PR and will also make another note to include you when I merge it.  Thanks again!

**Pull request closed without merging**